Snowflake Procedures Blog (#211)
ippontech/blog-usa: posts/a-primer-on-snowflake-stored-procedures.md

* Variable assignment
* [DESCRIBE](https://docs.snowflake.com/en/sql-reference/sql/desc.html) and [SHOW](https://docs.snowflake.com/en/sql-reference/sql/show.html) (See limitations on describe and show above.)

All other SQL statements cannot be called from inside an owner's rights stored procedure.

> If an owner’s rights stored procedure is called by a caller’s rights stored procedure, or vice-versa, the following rules apply:
> * A stored procedure behaves as a caller’s rights stored procedure if and only if the procedure and the entire call hierarchy above it are caller’s rights stored procedures.
> * An owner’s rights stored procedure always behaves as an owner’s rights stored procedure, no matter where it was called from.
> * Any stored procedure called directly or indirectly from an owner’s rights stored procedure behaves as an owner’s rights stored procedure.

##### Choosing between Owner's Rights and Caller's Rights
Create a stored procedure as an owner's rights stored procedure if **all** of the following is true:
* A task needs to be delegated to another user who will run with owner's privileges, not their own.

>   * Ex: If you need a user without DELETE privilege on a table to be able to call a procedure to delete old data, without manipulating current data.

Reviewer: ```suggestion
  * E.g., If you need a user without DELETE privilege on a table to be able to call a procedure to delete old data, without manipulating current data.
```

* ```RETURNS table()```
  Specifies the results returned by the stored procedure. This return value cannot be used since the call cannot be a part of an expression.
  For result_data_type, use the Snowflake data type that corresponds to the type of the language that you are using:
  * For JavaScript stored procedures, see [SQL and JavaScript Data Type Mapping](https://docs.snowflake.com/en/sql-reference/stored-procedures-javascript.html#label-stored-procedure-data-type-mapping).
  * For Snowflake Scripting, a [SQL data type]().
  * For Snowpark (Scala and Java) stored procedures, see:
    * [Appendix: Mapping Scala Data Types to Snowflake Data Types](https://docs.snowflake.com/en/sql-reference/stored-procedures-scala.html#label-sql-types-to-scala-types)
    * [SQL-Java Data Type Mappings for Parameters and Return Types](https://docs.snowflake.com/en/developer-guide/udf/java/udf-java-designing.html#label-sql-java-data-type-mappings).

* ```LANGUAGE SQL```

>   Specifies what language the stored procedure is written in.  A procedure can currently be written in one of the following languages: JavaScript, Snowflake Scripting (SQL), Scala (using Snowpark) and Java (using Snowpark) 

Reviewer: ```suggestion
  Specifies the language in which the stored procedure is written.  A procedure can currently be written in one of the following languages: JavaScript, Snowflake Scripting (SQL), Scala (using Snowpark), and Java (using Snowpark) 
```

---
# Access Control on Stored Procedures
To finish off our discussion of Stored Procedures, I am going to delve a little bit into the access control features Snowflake provides for Stored Procedures. There are two types of privileges that stored procedures utilize: Those directly on the procedure and those on the database objects that the procedure can access. Similar to other database objects in Snowflake, stored procedures are owned by a role and have one or more privileges that can be granted to other roles.

There are currently two privileges that can apply to stored procedures: USAGE and OWNERSHIP. For a role to use a stored procedure, it must either be the owner of the stored procedure or have been granted USAGE on the procedure ^[For a more in-depth view on Access Control in Snowflake, I refer you to [Access Control Considerations](https://docs.snowflake.net/manuals/user-guide/security-access-control-considerations.html) in the Snowflake Documentation.]. In addition, the roles executing the procedure must also have the USAGE privilege for all database objects accessed during the course of the procedure.

----

> During this blog, we've examined Snowflake Stored Procedures from creation to execution. We've also discussed some of the key differences between stored procedures and user-defined functions as well as examined session state and the two types of procedures Snowflake provides. To round out procedures, we briefly viewed access control in Snowflake and how privileges can affect procedure execution.

Reviewer: ```suggestion
During this blog, we have examined Snowflake Stored Procedures from creation to execution. We have also discussed some of the key differences between stored procedures and user-defined functions, as well as examined session state and the two types of procedures Snowflake provides. To round out procedures, we briefly viewed access control in Snowflake and how privileges can affect procedure execution.
```

----

> For more information on how Ippon Technologies, a Snowflake partner, can help your organization utilize the benefits of Snowflake for a migration from a traditional Data Warehouse, Data Lake or POC, contact sales@ipponusa.com.

Reviewer: ```suggestion
For more information on how Ippon Technologies, a Snowflake Services Partner, can help your organization utilize the benefits of Snowflake for a migration from a traditional data warehouse, data lake, or POC, contact sales@ipponusa.com.
```
